### PR TITLE
Add OpenVINO API 2.0 support

### PR DESCRIPTION
**Pull request by Semmer2**

### CODE REVIEW

File: patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch [hunk 1]
```diff
@@ -0,0 +1,1224 @@
...
+-dnn_suggest="libtensorflow libopenvino libtorch"
++dnn_suggest="libtensorflow libopenvino libopenvino2 libtorch"
+ dnn_deps="avformat swscale"
+ error_resilience_select="me_cmp"
+ faandct_deps="faan"
+@@ -6677,6 +6678,7 @@ enabled libopenjpeg       && { check_pkg_config libopenjpeg "libopenjp2 >= 2.1.0
+                                { require_pkg_config libopenjpeg "libopenjp2 >= 2.1.0" openjpeg.h opj_version -DOPJ_STATIC && add_cppflags -DOPJ_STATIC; } }
+ enabled libopenmpt        && require_pkg_config libopenmpt "libopenmpt >= 0.2.6557" libopenmpt/libopenmpt.h openmpt_module_create -lstdc++ && append libopenmpt_extralibs "-lstdc++"
+ enabled libopenvino       && require libopenvino c_api/ie_c_api.h ie_c_api_version -linference_engine_c_api
++enabled libopenvino2      && require_pkg_config libopenvino2 openvino openvino/c/openvino.h OK
```
Comment (xhaihao) [inline patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch:52]: Why is OK required here ?

File: patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch [hunk 2]
```diff
@@ -0,0 +1,1224 @@
...
++ * Lesser General Public License for more details.
++ *
++ * You should have received a copy of the GNU Lesser General Public
++ * License along with FFmpeg; if not, write to the Free Software
++ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
++ */
++
++/**
++ * @file
++ * DNN OpenVINO backend implementation.
```
Comment (xhaihao) [inline patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch:96]: s/OpenVINO/OpenVINO API v2 ?

File: patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch [hunk 3]
```diff
@@ -0,0 +1,1224 @@
+From 45324ace54135245680e59f313b604571aec2bd4 Mon Sep 17 00:00:00 2001
+From: Ting Fu <ting.fu@intel.com>
+Date: Thu, 22 Dec 2022 10:52:27 +0800
+Subject: [PATCH] lavfi/dnn: Add OpenVINO API 2.0 support
+
+OpenVINO API 2.0 was released in March 2022, which introduced new features.
+This commit implement current OpenVINO features with new 2.0 APIs. And will
+adding other features in API 2.0.
```
Comment (xhaihao) [inline patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch:8]: s/adding/add ?

File: patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch [hunk 4]
```diff
@@ -0,0 +1,1224 @@
+From 45324ace54135245680e59f313b604571aec2bd4 Mon Sep 17 00:00:00 2001
+From: Ting Fu <ting.fu@intel.com>
+Date: Thu, 22 Dec 2022 10:52:27 +0800
+Subject: [PATCH] lavfi/dnn: Add OpenVINO API 2.0 support
+
+OpenVINO API 2.0 was released in March 2022, which introduced new features.
+This commit implement current OpenVINO features with new 2.0 APIs. And will
```
Comment (xhaihao) [inline patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch:7]: s/implement/implements  ?

File: patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch [hunk 5]
```diff
@@ -0,0 +1,1224 @@
...
++    if (!header->nb_bboxes) {
++        return 0;
++    }
++
++    for (uint32_t i = 0; i < header->nb_bboxes; i++) {
++        bbox = av_get_detection_bbox(header, i);
++        if (bbox->x < 0 || bbox->w < 0 || bbox->x + bbox->w >= frame->width) {
++            return 0;
++        }
++        if (bbox->y < 0 || bbox->h < 0 || bbox->y + bbox->h >= frame->width) {
```
Comment (xhaihao) [inline patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch:303]: typo ? I guess it should be frame->height.

Comment (xhaihao) [reply patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch:303]: can bbox->w or bbox->h be 0 ?

File: patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch [hunk 6]
```diff
@@ -0,0 +1,1224 @@
...
++    ov_status_e status;
++    LastLevelTaskItem *lltask;
++    int ret = 0;
++    TaskItem *task;
++    OV2Context *ctx;
++    OV2Model *ov2_model;
++
++    if (ff_queue_size(inferenceq) == 0) {
++        ov_infer_request_free(request->infer_request);
++        av_freep(&request);
```
Comment (xhaihao) [inline patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch:188]: is there a same place to free such resources ?

File: patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch [hunk 7]
```diff
@@ -0,0 +1,1224 @@
...
++    }
++
++    request = ff_safe_queue_pop_front(ov2_model->request_queue);
++    if (!request) {
++        av_log(ctx, AV_LOG_ERROR, "unable to get infer request.\n");
++        ret = AVERROR(EINVAL);
++        goto err;
++    }
++
++    ret = execute_model_ov2(request, ov2_model->lltask_queue);
```
Comment (xhaihao) [inline patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch:618]: The return value from execute_model_ov2 is a dnn error code, however the expected value from get_output_ov2 is AVERROR.

File: patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch [hunk 8]
```diff
@@ -0,0 +1,1224 @@
...
++        while (ff_queue_size(ov2_model->lltask_queue) >= ctx->options.batch_size) {
++            request = ff_safe_queue_pop_front(ov2_model->request_queue);
++            if (!request) {
++                av_log(ctx, AV_LOG_ERROR, "unable to get infer request.\n");
++                return AVERROR(EINVAL);
++            }
++
++            ret = execute_model_ov2(request, ov2_model->lltask_queue);
++            if (ret != 0) {
++                return ret;
```
Comment (xhaihao) [inline patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch:745]: Please convert ret to an AVERROR.

File: patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch (hunk 5 above)
Comment (Semmer2) [reply patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch]: Yes, it should be `frame->height` and `bbox->w bbox->h` cannot be 0.

File: patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch (hunk 7 above)
Comment (Semmer2) [reply patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch:658]: Yes, this is an issue. I use a new func `ov2_map_error` to map OpenVINO error code to FFmpeg av_error, just like qsv did.

File: patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch (hunk 6 above)
Comment (Semmer2) [reply patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch:228]: Yes, it is also freed in the end of callback func `infer_completion_callback`.
And other resources free would be in `ff_dnn_free_model_ov2`.

File: patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch [hunk 9]
```diff
@@ -0,0 +1,1264 @@
...
++    task = lltask->task;
++    ov2_model = task->model;
++    ctx = &ov2_model->ctx;
++
++    if (task->async) {
++        ret = fill_model_input_ov2(ov2_model, request);
++        if (ret != 0) {
++            goto err;
++        }
++
```
Comment (xhaihao) [inline patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch:242]: You may move the above code (line 238 - line 241) out of if ... else ... block.

File: patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch [hunk 10]
```diff
@@ -0,0 +1,1264 @@
...
++    lltask = ff_queue_peek_front(inferenceq);
++    task = lltask->task;
++    ov2_model = task->model;
++    ctx = &ov2_model->ctx;
++
++    if (task->async) {
++        ret = fill_model_input_ov2(ov2_model, request);
++        if (ret != 0) {
++            goto err;
```
Comment (xhaihao) [inline patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch:240]: Is it needed to free request in the error path ?

Comment (Semmer2) [reply patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch]: Oh, I got you, add new free code in the end of this function.

File: patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch (hunk 9 above)
Comment (Semmer2) [reply patches/0083-lavfi-dnn-Add-OpenVINO-API-2.0-support.patch:241]: Sure, modified.

**Comment by xhaihao** [comment]
@Semmer2  this patch can't be applied to the latest FFmpeg, could you please fix your patch ?

**Comment by Semmer2** [comment]
Hi @xhaihao 
OpenVINO 2.0 patches rebased on latest cartwheel code. Please help to check, thank you.

Plus, there are conflicts applying patch 0033.

**Comment by Semmer2** [comment]
Hi @xhaihao 
OpenVINO 2.0 patches rebased on the latest cartwheel code, pleas help to check, thank you.

**Pull request merged**